resourcet/Control/Monad/Trans/Resource/Internal.hs

```diff
@@ -366,3 +368,44 @@ registerType istate rel = I.atomicModifyIORef istate $ \rm ->
             , ReleaseKey istate key
             )
         ReleaseMapClosed -> throw $ InvalidAccess "register'"
+
+-- | Thrown when one or more cleanup functions themselves throw an
+-- exception during cleanup.
+--
+-- @since 1.1.11
+data ResourceCleanupException = ResourceCleanupException !SomeException ![SomeException]
+  deriving (Show, Typeable)
+instance Exception ResourceCleanupException
+
+stateCleanupChecked :: ReleaseType -> I.IORef ReleaseMap -> IO ()
+stateCleanupChecked rtype istate = E.mask_ $ do
+    mm <- I.atomicModifyIORef istate $ \rm ->
+        case rm of
+            ReleaseMap nk rf m ->
+                let rf' = rf - 1
+                 in if rf' == minBound
+                        then (ReleaseMapClosed, Just m)
+                        else (ReleaseMap nk rf' m, Nothing)
+            ReleaseMapClosed -> throw $ InvalidAccess "stateCleanup"
+    case mm of
+        Just m -> do
+            res <- mapMaybeM (\x -> try (x rtype)) $ IntMap.elems m
+            case res of
+                [] -> return () -- nothing went wrong
+                e:es -> E.throwIO $ ResourceCleanupException e es
+        Nothing -> return ()
+  where
+    try :: IO () -> IO (Maybe SomeException)
+    try io = fmap (either Just (\() -> Nothing)) (E.try io)
+
+-- Note that this returns values in reverse order, which is what we
+-- want in the specific case of this function.
```

Comment:
Nitpick: Could name it `mapMaybeReverseM` then, I was thinking about exactly the order when reading the name at the call site.